File: lib/core/data/network/user/dto/user.dto.dart
import 'package:json_annotation/json_annotation.dart';

part 'user.dto.g.dart';

@JsonSerializable(explicitToJson: true)
class UserDto {
  final String id;
  final String? email;
  final String username;
  final String? avatarUrl;
  final String? bio;
  final String? role;
  final DateTime? createdAt;
  final UserCreditsDto? credits;
  final UserCountsDto? counts;
  final UserJobsPageDto? jobs;

  UserDto({
    required this.id,
    this.email,
    required this.username,
    this.avatarUrl,
    this.bio,
    this.role,
    this.createdAt,
    this.credits,
    this.counts,
    this.jobs
  });

  int get creditBalance => credits?.balance ?? 0;

  factory UserDto.fromJson(Map<String, dynamic> json) => _$UserDtoFromJson(json);
  Map<String, dynamic> toJson() => _$UserDtoToJson(this);
}

@JsonSerializable()
class UserCreditsDto {
  final int balance;
  final DateTime updatedAt;

  const UserCreditsDto({required this.balance, required this.updatedAt});

  factory UserCreditsDto.fromJson(Map<String, dynamic> json) =>
      _$UserCreditsDtoFromJson(json);

  Map<String, dynamic> toJson() => _$UserCreditsDtoToJson(this);
}

@JsonSerializable()
class UserCountsDto {
  final int followers;
  final int following;
  final int posts;
  final int jobs;

  UserCountsDto({
    this.followers = 0,
    this.following = 0,
    this.posts = 0,
    this.jobs = 0,
  });

  factory UserCountsDto.fromJson(Map<String, dynamic> json) =>
      _$UserCountsDtoFromJson(json);

  Map<String, dynamic> toJson() => _$UserCountsDtoToJson(this);
}

@JsonSerializable()
class UserJobDto {
  final String id;
  final String type;
  final String status;
  final int progress;
  final String prompt;
  final String? negativePrompt;
  final String modelName;
  final bool turboEnabled;
  final int creditCost;
  final String? provider;
  final String? errorMessage;
  final DateTime createdAt;
  final DateTime updatedAt;
  final DateTime? startedAt;
  final DateTime? completedAt;
  final DateTime? failedAt;

  const UserJobDto({
    required this.id,
    required this.type,
    required this.status,
    required this.progress,
    required this.prompt,
    this.negativePrompt,
    required this.modelName,
    required this.turboEnabled,
    required this.creditCost,
    this.provider,
    this.errorMessage,
    required this.createdAt,
    required this.updatedAt,
    this.startedAt,
    this.completedAt,
    this.failedAt,
  });

  factory UserJobDto.fromJson(Map<String, dynamic> json) =>
      _$UserJobDtoFromJson(json);

  Map<String, dynamic> toJson() => _$UserJobDtoToJson(this);
}

@JsonSerializable(explicitToJson: true)
class UserJobsPageDto {
  final List<UserJobDto> data;
  final String? nextCursor;
  final int take;

  const UserJobsPageDto({
    required this.data,
    this.nextCursor,
    required this.take,
  });

  factory UserJobsPageDto.fromJson(Map<String, dynamic> json) =>
      _$UserJobsPageDtoFromJson(json);

  Map<String, dynamic> toJson() => _$UserJobsPageDtoToJson(this);
}

File: composeApp/src/commonMain/kotlin/ie/app/neuragen/data/network/model/UserDto.kt
package ie.app.neuragen.data.network.model

import kotlinx.serialization.Serializable

@Serializable
data class UserUpdateDto(
    val username: String? = null,
    val bio: String? = null,
    val avatarUrl: String? = null
)

@Serializable
data class UserDto(
    val id: String,
    val email: String,
    val googleId: String? = null,
    val password: String? = null,
    val username: String,
    val bio: String? = null,
    val avatarUrl: String? = null,
    val role: String, // FREE|PRO|ADMIN
    val proExpiresAt: String? = null,
    val createdAt: String
)

@Serializable
data class UserMeDto(
    val id: String,
    val email: String,
    val username: String,
    val avatarUrl: String? = null,
    val bio: String? = null,
    val role: String,
    val proExpiresAt: String? = null,
    val createdAt: String,
    val credits: CreditBalanceDto,
    val counts: UserCountsDto,
    val jobs: JobsPaginationDto
)

@Serializable
data class CreditBalanceDto(
    val balance: Int,
    val updatedAt: String
)

@Serializable
data class UserCountsDto(
    val followers: Int,
    val following: Int,
    val posts: Int,
    val jobs: Int
)

@Serializable
data class CreditTopupRequest(
    val amount: Int,
    val note: String? = null
)

@Serializable
data class CreditTopupResponse(
    val userId: String,
    val amount: Int,
    val balance: Int,
    val reason: String,
    val transactionId: String,
    val note: String? = null,
    val createdAt: String
)

@Serializable
data class UserPublicDto(
    val id: String,
    val username: String,
    val avatarUrl: String? = null,
    val bio: String? = null,
    val credits: CreditBalanceDto? = null
)
